code/ai/aigoals.cpp

```diff
@@ -1227,9 +1228,10 @@ int ai_find_goal_index( ai_goal* aigp, int mode, int submode, int priority )
 
 /* Remove a goal from the given goals structure
  * Returns the index of the goal that it clears out.
- * This is important so that if active_goal == index you can set AI_GOAL_NONE
+ * This is important so that if active_goal == index you can set AI_GOAL_NONE.
+ * NOTE: Callers should check the value of remove_more.  If it is true, the function should be called again.
  */
-int ai_remove_goal_sexp_sub( int sexp, ai_goal* aigp )
+int ai_remove_goal_sexp_sub( int sexp, ai_goal* aigp, bool &remove_more )
 {
```

Comment:
Perhaps I'm missing something, but I'm not sure what the advantage is of making the caller responsible for looping. `bool remove_all = false` would remove that requirement and maintain API compatibility. (That said, this function is _almost_ a misplaced implementation detail, so if it makes the implementation easier or clearer, that's probably fine.)

Reply:
It's definitely implementation; I spent a bit of time puzzling out the design before settling on the parameter.  Turning `ai_remove_goal_sexp_sub()` into a loop would be inelegant, but the real kicker is that any given loop iteration might need to deactivate the active goal, and this requires information that only the caller has.